### ISSUE TITLE
workSpaceSize check assertion fails in ConvolutionBackwardWeights()  - DEBUG builds only

### ISSUE THREAD

**Issue by syoyo** [ROCmSoftwarePlatform/MIOpen#72]
## System

* Ubuntu 18.04
* ROCm 2.0
* MIOpen `master` branch + Debug build
* VEGA56(gfx900)

I have encountered assertion failure when trying to run keithito's tacotron.

https://github.com/keithito/tacotron

## How to reproduce

Download LJ-Speech data and setup as described in Tacotron's README.

Set `MIOPEN_ENABLE_LOGGING=1` and `MIOPEN_LOG_LEVEL=6`

Then, run train.py

```
$ gdb python
$ set args train.py
...
python: /home/syoyo/work/MIOpen/src/ocl/convolutionocl.cpp:4111: auto miopen::ConvolutionDescriptor::ConvolutionBackwardWeights(miopen::Handle &, const void *, const miopen::TensorDescriptor &, ConstData_t, const miopen::TensorDescriptor &, ConstData_t, miopenConvBwdWeightsAlgorithm_t, const void *, const miopen::TensorDescriptor &, Data_t, Data_t, size_t)::(anonymous class)::operator()(type-parameter-0-0) const: Assertion `(workSpace != nullptr) && (workSpaceSize >= workSpaceSizeDirect)' failed.
Thread 47 "python" received signal SIGABRT, Aborted.
...
$1 = (Data_t &) @0x7fff067f8f98: 0x60f235400
(gdb) p workSpaceSize
$2 = (size_t &) @0x7fff067f8fa0: 1048576
(gdb) p workSpaceSizeDirect
$3 = 4194304
```
Full log: [issue_72_full_log.txt](https://github.com/ROCmSoftwarePlatform/MIOpen/files/2760145/issue_72_full_log.txt)

`workSpaceSizeDirect` is 4x larger(due to sizeof(float)?) than `workSpaceSize`.

I have modified MIOpen source code to print the value of `BackwardWeightsGetWorkSpaceSizeDirect`

**Comment by atamazov** [comment]
@syoyo The patch which shall resolve the issue resides [here](https://github.com/ROCmSoftwarePlatform/MIOpen/issues/70#issuecomment-454457115). Please close the issue if it fixes the problem. Thanks.

**Comment by syoyo** [comment]
@atamazov At least the patch now avoids assertion failure, but an application now goes into infinite loop of compiling MIOpen conv kernels(by glimpsing MIOpen logs, it looks it infinitely try to search conv kernel)

**Comment by atamazov** [comment]
@syoyo Theoretically, this could be not an infinite loop but auto-tuning (exhaustive search) process. Please check if [autotuning is enabled in the environment](../blob/master/doc/src/perfdatabase.md#auto-tuning-the-kernels).

Anyway, I am unable to reproduce the "loop" problem with this config (`./bin/MIOpenDriver conv -n 32 -c 128 -H 1 -W 111 -k 128 -c 128 -y 1 -x 2 -p 0 -q 0 -V 0 -t 1 -w 1`). Please attach log. Thanks.

**Comment by syoyo** [comment]
>  I am unable to reproduce the "loop" problem with this config (./bin/MIOpenDriver conv -n 32 -c 128 -H 1 -W 111 -k 128 -c 128 -y 1 -x 2 -p 0 -q 0 -V 0 -t 1 -w 1). 

Ah, I meant "loop" occurs when I run tacotron app, not `MIOpenDriver` solely.
I'll also look into autotuning document.

Currently I'm running a learning task for #70 regression. I'll attach log of #72 loop issue after that(it may take half or one day)

**Comment by atamazov** [comment]
@syoyo I highly recommend opening another issue for the loop problem. Let's keep this issue dedicated to workspace size check assertion. Thanks.

**Comment by syoyo** [comment]
@atamazov I understand. I'll open another issue once ready to submit log file > I highly recommend opening another issue for the loop problem

So this issue is ready to clone once PR https://github.com/ROCmSoftwarePlatform/MIOpen/pull/73 has been merged(Inherently its @atamazov's patch)

**Comment by atamazov** [comment]
I am not sure that #73 will be merged. AFAIK we are going to release 1.7.1 instead, which shall contain a set of fixes & improvements, including fix for #70 and #72.

**Comment by atamazov** [comment]
Note that the issue only occurs with Debug builds of MIOpen (i.e. those which have assertions enabled).

**Comment by syoyo** [comment]
Fix has been included in 1.7.1: https://github.com/ROCmSoftwarePlatform/MIOpen/commit/a478ac8d5d3896e910721832379c16ea5ced06bc
